[ML] Include categorizer size in model size stats

droberts195 opened the pull request:
This change adds the size of the categorizer used in categorization
jobs to the model_bytes field reported in model_size_stats, making
it a more accurate reflection of actual memory usage.

As with anomaly detectors, only the model library classes are
instrumented, which means that the small amount of additional api
library memory usage is not included.

Closes #724

droberts195 commented on lib/api/unittest/CSingleStreamDataAdderTest.cc:
```diff
@@ -73,42 +73,42 @@ void detectorPersistHelper(const std::string& configFileName,
-        JOB_ID, limits, fieldConfig, modelConfig, wrappedOutputStream,
-        std::bind(&reportPersistComplete, std::placeholders::_1,
-                  std::ref(origSnapshotId), std::ref(numOrigDocs)),
-        nullptr, -1, "time", timeFormat);
```
It looks like there are masses of changes in this file but most are whitespace differences caused by me moving braces to adjust object lifetimes.  I did this after changing the resource monitor to reduce memory consumption on unregistering a component.  This test was relying on the old (unintuitive in my opinion) behaviour.

Best to review this file with "ignore whitespace differences" enabled!

edsavage replied:
👍

edsavage commented on lib/api/CFieldDataCategorizer.cc:
```diff
@@ -159,12 +162,19 @@ int CFieldDataCategorizer::computeCategory(const TStrStrUMap& dataRowFields) {
 
-    bool exampleAdded = m_ExamplesCollector.add(categoryId, fieldValue);
-    bool searchTermsChanged = this->createReverseSearch(categoryId);
+    bool exampleAdded{m_DataCategorizer->addExample(categoryId, fieldValue)};
```
Just realised that while we have a comment in `STYLEGUIDE.md` for (non-)boolean accessor functions:

```Boolean accessor functions MUST be named as isComplete NOT complete```

we don't seem to have anything similar for mutator functions. Should there be? If so what?

droberts195 replied:
In the old coding standards Word document the section titled "Non-boolean accessor functions MUST be named as 'clientId' NOT 'getClientId'" had a detailed description of "Accessor functions MUST NOT be prefixed with get or set."

I think people are still adhering to setter methods not being prefixed with set.

But this is not a setter method, it's adding an item to a list.  So I think it's better that it does contain a verb to indicate that it's not a setter method.  If it was just `example` then people might assume it was a setter method setting a single example.

edsavage commented on lib/api/CFieldDataCategorizer.cc:
```diff
@@ -159,12 +162,19 @@ int CFieldDataCategorizer::computeCategory(const TStrStrUMap& dataRowFields) {
 
-    bool exampleAdded = m_ExamplesCollector.add(categoryId, fieldValue);
-    bool searchTermsChanged = this->createReverseSearch(categoryId);
+    bool exampleAdded{m_DataCategorizer->addExample(categoryId, fieldValue)};
```
```suggestion
    bool addExample{m_DataCategorizer->addExample(categoryId, fieldValue)};
```

edsavage commented on lib/api/CFieldDataCategorizer.cc:
```diff
@@ -298,12 +308,7 @@ bool CFieldDataCategorizer::acceptRestoreTraverser(core::CStateRestoreTraverser&
-            traverser.haveBadState()) {
-            LOG_ERROR(<< "Cannot restore categorizer, unexpected element: "
-                      << traverser.value());
+        if (m_DataCategorizer->restoreExamplesCollector(traverser) == false) {
```
👍  💯

edsavage commented on lib/api/CFieldDataCategorizer.cc:
```diff
@@ -410,6 +415,9 @@ void CFieldDataCategorizer::acceptPersistInserter(
 bool CFieldDataCategorizer::periodicPersistStateInBackground() {
     LOG_DEBUG(<< "Periodic persist categorizer state");
 
+    // Make sure model size stats are up to date
```
```suggestion
    // Make sure that the model size stats are up to date
```

edsavage commented on lib/api/unittest/CPersistenceManagerTest.cc:
```diff
@@ -162,8 +162,8 @@ class CTestFixture {
 
-        // Replace the zero byte separators so the expected/actual strings don't get
-        // truncated by CppUnit if the test fails
+        // Replace the zero byte separators to avoid '\0's in the output if the
```
👍

edsavage commented on lib/model/CAnomalyDetector.cc:
```diff
@@ -173,11 +171,6 @@ void CAnomalyDetector::zeroModelsToTime(core_t::TTime time) {
-    // As the model pointer will change during restore, we unregister
-    // the detector from the resource monitor. We can register it
-    // again at the end of restore.
-    m_Limits.resourceMonitor().unRegisterComponent(*this);
```
👍

edsavage commented on lib/model/CBaseTokenListDataCategorizer.cc:
```diff
@@ -31,19 +31,18 @@ const std::string TOKEN_TAG("a");
 }
 
 CBaseTokenListDataCategorizer::CBaseTokenListDataCategorizer(
+    CLimits& limits,
```
```suggestion
CTokenListDataCategorizerBase::CTokenListDataCategorizerBase(
    CLimits& limits,
```

edsavage commented on lib/model/CCategoryExamplesCollector.cc:
```diff
@@ -143,6 +144,18 @@ void CCategoryExamplesCollector::clear() {
+    core::CMemoryDebug::dynamicSize("m_ExamplesByCategory", m_ExamplesByCategory, mem);
+}
+
+//! Get the memory used by this examples collector.
```
```suggestion
//! Return the memory used by this examples collector.
```

droberts195 replied:
This doxygen comment shouldn't have been in the `.cc` file at all.  I have removed it.

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -45,20 +44,22 @@ void CResourceMonitor::memoryUsageReporter(const TMemoryUsageReporterFunc& repor
-    if (itr == m_Detectors.end()) {
-        LOG_ERROR(<< "Inconsistency - component has not been registered: " << &detector);
+void CResourceMonitor::unRegisterComponent(CMonitoredResource& resource) {
+    auto itr = m_Resources.find(&resource);
```
```suggestion
LOG_TRACE(<< "Unregistering component: " << &resource);
    auto itr = m_Resources.find(&resource);
```

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -45,20 +44,22 @@ void CResourceMonitor::memoryUsageReporter(const TMemoryUsageReporterFunc& repor
+void CResourceMonitor::unRegisterComponent(CMonitoredResource& resource) {
+    auto itr = m_Resources.find(&resource);
+    if (itr == m_Resources.end()) {
+        LOG_ERROR(<< "Inconsistency - component has not been registered: " << &resource);
```
```suggestion
        LOG_ERROR(<< "Inconsistency. - Cannot unregister a component that has not yet been registered: " << &resource);
```

droberts195 replied:
Reformatting like this would make the message inconsistent with all the other similar ones - `grep Inconsistency lib/*/*.cc` and you'll see.  Also, these are basically soft asserts, i.e. things that should never happen but if they did we decided not to abort the program but let it carry on.  So hopefully nobody will ever see such a message logged and the exact wording is not particularly important.

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -45,20 +44,22 @@ void CResourceMonitor::memoryUsageReporter(const TMemoryUsageReporterFunc& repor
 
-    LOG_TRACE(<< "Unregistering component: " << &detector);
-    m_Detectors.erase(itr);
+    LOG_TRACE(<< "Unregistering component: " << &resource);
```
```suggestion
```

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -45,20 +45,22 @@ void CResourceMonitor::memoryUsageReporter(const TMemoryUsageReporterFunc& repor
-    LOG_TRACE(<< "Unregistering component: " << &detector);
-    m_Detectors.erase(itr);
+    LOG_TRACE(<< "Unregistering component: " << &resource);
+    m_CurrentMonitoredResourceMemory -= itr->second;
```
I bet :-/

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -205,16 +213,16 @@ std::size_t CResourceMonitor::allocationLimit() const {
+void CResourceMonitor::memUsage(CMonitoredResource* resource) {
+    auto itr = m_Resources.find(resource);
+    if (itr == m_Resources.end()) {
+        LOG_ERROR(<< "Inconsistency - component has not been registered: " << resource);
```
```suggestion
        LOG_ERROR(<< "Inconsistency - component has not yet been registered: " << resource);
```

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -205,16 +213,16 @@ std::size_t CResourceMonitor::allocationLimit() const {
     std::size_t modelCurrentUsage = core::CMemory::dynamicSize(itr->first);
     itr->second = modelCurrentUsage;
-    m_CurrentAnomalyDetectorMemory += (modelCurrentUsage - modelPreviousUsage);
+    m_CurrentMonitoredResourceMemory += (modelCurrentUsage - modelPreviousUsage);
```
```suggestion
    m_MonitoredResourceCurrentMemory += (modelCurrentUsage - modelPreviousUsage);
```

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -232,8 +240,8 @@ bool CResourceMonitor::needToSendReport() {
     if (!m_AllocationFailures.empty()) {
-        core_t::TTime lastestAllocationError = (--m_AllocationFailures.end())->first;
-        if (lastestAllocationError > m_LastAllocationFailureReport) {
+        core_t::TTime latestAllocationError = (--m_AllocationFailures.end())->first;
```
```suggestion
        core_t::TTime latestAllocationError{--m_AllocationFailures.end())->first}
```

edsavage commented on lib/model/CResourceMonitor.cc:
```diff
@@ -346,7 +351,7 @@ std::size_t CResourceMonitor::lowLimit() const {
 
 std::size_t CResourceMonitor::totalMemory() const {
-    return m_CurrentAnomalyDetectorMemory + m_ExtraMemory +
+    return m_CurrentMonitoredResourceMemory + m_ExtraMemory +
```
```suggestion
    return m_MonitoredResourceCurrentMemory + m_ExtraMemory +
```

edsavage commented on lib/model/unittest/CResourceMonitorTest.cc:
```diff
@@ -21,6 +23,22 @@ BOOST_AUTO_TEST_SUITE(CResourceMonitorTest)
 
+namespace {
+using TTokenListDataCategorizerKeepsFields =
+    model::CTokenListDataCategorizer<true,  // Warping
```
named constants would be better here I think...

droberts195 replied:
Actually since in this test it doesn't matter exactly how the categorizer is configured I switched it to use the default template arguments in this file.  Then this whole unnamed namespace goes.

edsavage commented on lib/model/unittest/CResourceMonitorTest.cc:
```diff
@@ -325,7 +366,7 @@ BOOST_FIXTURE_TEST_CASE(testMonitor, CTestFixture) {
 
         // Go up to 10 bytes, triggering a need
-        mon.m_CurrentAnomalyDetectorMemory = 10;
+        mon.m_CurrentMonitoredResourceMemory = 10;
```
```suggestion
        mon.m_MonitoredResourceCurrentMemory = 10;
```

and elsewhere below...

edsavage commented on include/api/CFieldDataCategorizer.h:
```diff
@@ -78,7 +76,7 @@ class API_EXPORT CFieldDataCategorizer : public CDataProcessor {
     CFieldDataCategorizer(const std::string& jobId,
                           const CFieldConfig& config,
-                          const model::CLimits& limits,
+                          model::CLimits& limits,
```
I thought there was something in `STYLEGUIDE.md` about parameter ordering based on type but I can't find it.. Maybe it was in the Prelert guide? I'll check later.

droberts195 replied:
Yes:

> Function parameters MUST be ordered as [in] [in,out] [out]

Nobody else takes any notice of that any more so there are masses of violations of it which is why I didn't bother here, but since you noticed I'll change this one.

droberts195 replied:
Actually, come to think of it this one is flukily already in the mandated order, because `limits` is effectively now an `[in,out]` variable.

droberts195 commented on include/model/CResourceMonitor.h:
```diff
@@ -53,8 +52,8 @@ class MODEL_EXPORT CResourceMonitor {
 public:
-    using TDetectorPtrSizePr = std::pair<CAnomalyDetector*, std::size_t>;
-    using TDetectorPtrSizeUMap = boost::unordered_map<CAnomalyDetector*, std::size_t>;
+    using TResourcePtrSizePr = std::pair<CMonitoredResource*, std::size_t>;
```
When I used CLion's refactoring functionality on this line it revealed that this particular typedef isn't used anywhere, so I completely removed it - it must be dead code left over from years ago.  But I renamed the one on the line below as per this suggestion.

droberts195 commented on include/model/CResourceMonitor.h:
```diff
@@ -53,8 +52,8 @@ class MODEL_EXPORT CResourceMonitor {
 public:
-    using TDetectorPtrSizePr = std::pair<CAnomalyDetector*, std::size_t>;
-    using TDetectorPtrSizeUMap = boost::unordered_map<CAnomalyDetector*, std::size_t>;
+    using TResourcePtrSizePr = std::pair<CMonitoredResource*, std::size_t>;
```
When I used CLion's refactoring functionality on this line it revealed that this particular typedef isn't used anywhere, so I completely removed it - it must be dead code left over from years ago.  But I renamed the one on the line below as per this suggestion.

droberts195 commented on include/model/CMonitoredResource.h:
```diff
@@ -0,0 +1,68 @@
+//! pruning.  Methods related to pruning have default implementations
+//! consistent with this assumption.
+//!
+class MODEL_EXPORT CMonitoredResource {
```
Historically we've only used `Base` or `Interface` in abstract base class names where they were introduced a long time after an original single concrete implementation, that concrete implementation had already taken the obvious name, and it was considered confusing to change the name of the original class.  Appending `Base` or `Interface` to abstract base class names is not something that's commonly done in the codebase.

edsavage replied:
Discussed internally and decided against this suggestion

droberts195 commented:
I think I addressed most of the review points in 49c102c03df6cd77a95b47e82ab9be7eed8b55ac or replies where I didn't make a change.  The review had multiple conflicting suggestions for the same comment in a couple of places, so obviously I couldn't exactly match both suggestions, but I did change those comments.

droberts195 commented:
retest
